### PR TITLE
[flang] Simplify hlfir.sum total reductions.

### PR DESCRIPTION
**Pull request by vzakhari**
I am trying to switch to keeping the reduction value in a temporary
scalar location so that I can use hlfir::genLoopNest easily.
This also allows using omp.loop_nest with worksharing for OpenMP.

### CODE REVIEW

File: flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp [hunk 1]
```diff
@@ -141,42 +155,32 @@ class SumAsElementalConversion : public mlir::OpRewritePattern<hlfir::SumOp> {
...
-      // for the DIM dimension.
-      mlir::Value iv = doLoop.getInductionVar();
-      llvm::SmallVector<mlir::Value> indices{inputIndices};
-      indices.insert(indices.begin() + dimVal - 1, iv);
-
-      mlir::OpBuilder::InsertionGuard guard(builder);
-      builder.setInsertionPointToStart(doLoop.getBody());
-      mlir::Value reductionValue = doLoop.getRegionIterArgs()[0];
+      bool emitWorkshareLoop =
+          isTotalReduction ? flangomp::shouldUseWorkshareLowering(sum) : false;
```
Comment (tblah) [inline flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp:169]: We cannot use the workshare lowering here. This lowering allows the loop iterations to be split over multiple threads. The updates to the reduction value are not atomic so it is not safe to update it from these multiple threads.

OpenMP has a reduction clause which would handle the reduction variable updates correctly, but doing that is outside the scope of this patch.

Comment (vzakhari) [reply flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp:169]: Thank you, Tom! Yes, this won't work without a reduction clause.  I will change the code and put a TODO note about this.

File: flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp [hunk 2]
```diff
@@ -105,34 +105,47 @@ class SumAsElementalConversion : public mlir::OpRewritePattern<hlfir::SumOp> {
...
-      // TODO: produce the slices and new SUM after adding a pattern
-      // for expanding total reduction SUM case.
-      mlir::Type indexType = builder.getIndexType();
-      auto one = builder.createIntegerConstant(loc, indexType, 1);
-      auto ub = builder.createConvert(loc, indexType, dimExtent);
+      // If DIM is not present, do total reduction.
 
+      // Create temporary scalar for keeping the running reduction value.
+      mlir::Value reductionTemp =
+          builder.createTemporaryAlloc(loc, elementType, ".sum.reduction");
```
Comment (vzakhari) [inline flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp:145]: @jeanPerier, what do you think about calling this outside of `genKernel`?  It looks like it results in stacksave/stackrestore in the stack reclaim pass (after the elemental is transformed into loops), which is not ideal. I think it should be safe to hoist this call provided that the initializing store is kept inside the elemental.

Comment (jeanPerier) [reply flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp:145]: Makes some sense to me, the only impact I see is that it may make harder parallelization of SUM(, DIM) which is otherwise trivial (each threads does reduces into an element of the result array).

Can you try what happens with a SUM(, DIM) inside a workshare construct? Since you enable the rewrite to an elemental, I think that the elemental to omp loop should kick in and hoisting the alloca may be bad there.

Maybe the stack reclaim pass should hoist constant size alloca outside of loops (with the assumption that parallelization of the loops happened at that point), at least for scalars. This may have impacts on the stack size of course, but for scalars that should be limited.

Since SUM(, DIM) was not parallelized before anyway, your solution would still be acceptable to me though.

**Review by jeanPerier — APPROVED** [review]
LGTM

File: flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp (hunk 2 above)
Comment (tblah) [reply flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp:145]: `fir::FirOpBuilder::getAllocaBlock` understands OpenMP operations and should give safe insertion points to move allocas in the stack reclaim pass. OpenMP parallelisation will all have happened by then.

If OpenMP workshare support is blocking optimizations in earlier passes please let me know and I will see if I can rethink the design.

**Review by tblah — APPROVED** [review]
Thanks for the fix!

**Comment by vzakhari** [comment]
I tried the following test:
```
  subroutine test(x, y)
    real :: x(:,:), y(:)
    !$omp parallel workshare
    y = sum(x,dim=1)
    !$omp end parallel workshare
  end subroutine test
```

When I place the alloca inside the elemental, I get the following:
```
// -----// IR Dump After OptimizedBufferization (opt-bufferization) //----- //
  omp.parallel {
    omp.workshare {
      omp.workshare.loop_wrapper {
        omp.loop_nest (%arg2) : index = (%c1) to (%4#1) inclusive step (%c1) {
          %5 = fir.alloca f32 {bindc_name = ".sum.reduction", pinned}
          fir.do_loop

// -----// IR Dump After LowerWorkshare (lower-workshare) //----- //
    omp.parallel {
      omp.wsloop nowait {
        omp.loop_nest (%arg2) : index = (%c1) to (%6#1) inclusive step (%c1) {
          %7 = fir.alloca f32 {bindc_name = ".sum.reduction", pinned}
          fir.do_loop

// -----// IR Dump Before LLVMAddComdats (llvm-add-comdats) //----- //
    omp.parallel {
      %52 = llvm.alloca %51 x f32 {bindc_name = ".sum.reduction", pinned} : (i64) -> !llvm.ptr
      omp.wsloop nowait {
        omp.loop_nest (%arg2) : i64 = (%5) to (%62) inclusive step (%5) {
          llvm.store %4, %52 {tbaa = [#tbaa_tag3]} : f32, !llvm.ptr

; *** IR Dump After Annotation2MetadataPass on [module] ***
define internal void @_QFPtest..omp_par(ptr noalias %tid.addr, ptr noalias %zero.addr, ptr %0) #1 {
omp.par.region1:                                  ; preds = %omp.par.region
  %2 = alloca float, i64 1, align 4
...
define internal void @_QFPtest(ptr %0, ptr %1) #0 {
  call void (ptr, i32, ptr, ...) @__kmpc_fork_call(ptr @1, i32 1, ptr @_QFPtest..omp_par, ptr %structArg)
```

So it seems that we end up allocating the temporary in each thread, and then the stack is automatically reclaimed after exiting `_QFPtest..omp_par`. It also looks fine to me at the MLIR level.

When I place the alloca outside the elemental:
```
// -----// IR Dump After OptimizedBufferization (opt-bufferization) //----- //
  omp.parallel {
    omp.workshare {
      %3 = fir.alloca f32 {bindc_name = ".sum.reduction", pinned}
      omp.workshare.loop_wrapper {
        omp.loop_nest (%arg2) : index = (%c1) to (%5#1) inclusive step (%c1) {
          fir.store %cst to %3 : !fir.ref<f32>
          fir.do_loop

// -----// IR Dump After LowerWorkshare (lower-workshare) //----- //
    omp.parallel {
      %5 = fir.alloca f32 {bindc_name = ".sum.reduction", pinned}
      omp.single copyprivate(%5 -> @_workshare_copy_f32 : !fir.ref<f32>) {
        omp.terminator
      }
      omp.wsloop nowait {
        omp.loop_nest (%arg2) : index = (%c1) to (%7#1) inclusive step (%c1) {
          fir.store %cst to %5 : !fir.ref<f32>
          fir.do_loop

// -----// IR Dump Before LLVMAddComdats (llvm-add-comdats) //----- //
    omp.parallel {
      %52 = llvm.alloca %51 x f32 {bindc_name = ".sum.reduction", pinned} : (i64) -> !llvm.ptr
      omp.single copyprivate(%52 -> @_workshare_copy_f32 : !llvm.ptr) {
        omp.terminator
      }
      omp.wsloop nowait {
        omp.loop_nest (%arg2) : i64 = (%5) to (%63) inclusive step (%5) {
          llvm.store %4, %52 {tbaa = [#tbaa_tag4]} : f32, !llvm.ptr

; *** IR Dump After Annotation2MetadataPass on [module] ***
define internal void @_QFPtest..omp_par(ptr noalias %tid.addr, ptr noalias %zero.addr, ptr %0) #1 {
omp.par.region1:                                  ; preds = %omp.par.region
  %2 = alloca float, i64 1, align 4
  %3 = alloca i32, align 4
  store i32 0, ptr %3, align 4
  %omp_global_thread_num2 = call i32 @__kmpc_global_thread_num(ptr @1)
  %4 = call i32 @__kmpc_single(ptr @1, i32 %omp_global_thread_num2)
  %5 = icmp ne i32 %4, 0
  br i1 %5, label %omp_region.body, label %omp_region.end

omp_region.end:                                   ; preds = %omp.par.region1, %omp.region.cont3
  %omp_global_thread_num4 = call i32 @__kmpc_global_thread_num(ptr @1)
  %6 = load i32, ptr %3, align 4
  call void @__kmpc_copyprivate(ptr @1, i32 %omp_global_thread_num4, i64 0, ptr %2, ptr @_workshare_copy_f32, i32 %6)
...
define internal void @_QFPtest(ptr %0, ptr %1) #0 {
  call void (ptr, i32, ptr, ...) @__kmpc_fork_call(ptr @1, i32 1, ptr @_QFPtest..omp_par, ptr %structArg)
```

So besides the `single copyprivate` we end up with the same behavior, I think.


It seems to me hoisting the alloca should be okay, but might be not enough to get rid of stacksave/stackrestore always (e.g. if we generate the elemental inside another loop, we will end up with the stack bookkeeping anyway).

I am now inclining to get back to the SSA reduction, though modifying `hlfir::genLoopNest` for this will look awkward (due to no support of SSA reduction by the related OpenMP constructs). So I will probably add something like `hlfir::genLoopNestWithSSAReduction` that will only generate `fir.do_loop`s, and then we can try to merge the two while also resolving it for OpenMP.

What do you think?

**Comment by tblah** [comment]
> What do you think?

Thanks for taking the time to take an in-depth look at OpenMP. 

I am worried the added `single copyprivate` could create a performance problem. Although the real OpenMP applications I have looked at so far do not use any transformational intrinsics inside of `workshare`.

I agree that `hlfir::genLoopNestWithSSAReduction` will be a good idea, and we can (eventually) use a custom implementation for openmp using the `omp.wsloop reduction(...)`. 

Pinging @ivanradanov and @Thirumalai-Shaktivel from AMD

**Comment by vzakhari** [comment]
I added `genLoopNestWithReductions` in the latest commit. It actually allows using either SSA or in-memory reductions, and the implementation of `genLoopNestWithReductions` may choose one or another depending on the context. This requires that the innermost loop body generator is passed with the initial values of the iter-args (reductions in this case), and returns the updated ("incremeneted") values, so that `genLoopNestWithReductions` can generate proper update of the reduction SSA/in-memory values depending on what the actual loop operations are generated. So the body generator does not need to be aware about what terminators the particular loop operations are using and how reduction values are updated.

I think `genLoopNestWithReductions` is more generic than `genLoopNest`, but I did not want to replace all uses of `genLoopNest` right now, moreover, `genLoopNestWithReductions` does not support generating proper worksharing operations right now.

**Review by tblah — APPROVED** [review]
This looks great to me. I don't want to hold this up on OpenMP concerns.

I can roughly imagine how this could be implemented for OpenMP (not for this patch). I think ultimately we will need the genBody callback to operate on scalars so that we can generate the right reduction clause implementation (or alternatively have an argument saying what the openmp intrinsic reduction kind is then use the lowering code to generate the `omp.declare_reduction`).

I'm imagining something like
```
omp.declare_reduction @something : type init {
  ^bb0(%unused: type)
    %init_val = // The arith.constant passed as the init value for the reduction variable
    omp.yield(%init_val)
} combiner {
^bb0(%lhs: type, %rhs: type):
  // generated by the genBody callback. e.g. for type == i32
  %res = arith.addi %lhs, %rhs : i32
  omp.yield(%res)
}

func.func [...] {
  %fortran_variable:2 = hlfir.declare [...]
  omp.wsloop reduction(@something %fortran_variable#0 -> %arg0 : !fir.ref<type>) {
    omp.loop_nest /*indices are %arg1...%argn*/ {
      %privatized_variable = hlfir.declare %arg0 [...]
      // generated by genLoopWIthReductions:
      %rhs = hlfir.designate [...]
      // generated by genBody:
      %res = operation %privatized_variable %rhs : type
      omp.yield
    }
  }
  // The result has been stored to %fortran_variable
}
```
This would give some loss of generality on the indexing. Maybe there should be two callbacks: one for indexing and one for the "combiner" op. IMO this can be done in a later patch.

**Comment by vzakhari** [comment]
Thank you for the example, Tom! Yes, it should be possible to split the body generator into parts so that it works for generating the combiner code and the rhs value code.
